### ISSUE TITLE
Add `traverseM` to Traversable to recover short-circuiting

### ISSUE THREAD

**Issue by stephentetley** [flix/flix#4540]
Currently `Traversable` supports `traverse` as per its Haskell counterpoint. `traverse` in both Haskell and Flix is applicative, however I just discovered that lazy evaluation makes Haskell's `traverse` different to Flix's `traverse`.

In a nutshell lazy evaluation lets Haskell's `traverse` short-circuit the traversal if the mapping function `f` supplied to `traverse` fails (for applicatives that support failure, of course). I don't think we can recover this in a strict language - the essence of applicative is that control flow is static so we can't choose to stop evaluating the remaining elements if one fails. However in a lazy language we get short-circuiting because we don't demand the remaining elements after a failure - this is a property of lazyness and not a contradiction of applicative.

In a strict language we can recover short circuiting for monads with failure (rather than applicatives) if we write a version of traverse that uses `bind` / `flatMap` to inspect the result of applying the mapping function `f` at each step and stop the traversal on the first failure. The implementation for List looks like this:

~~~

use Monad.{>>=};

///
/// The implementation of `flatMap` for the Monad instance must be stack safe.
/// If the Monad instance supports failure then the traversal short circuits on the first failure.
///
pub def traverseM(f: a -> m[b] \ ef, l: List[a]): m[List[b]] \ ef with Monad[m] = 
    def loop(ll, k) = match ll {
        case Nil => k(Nil)
        case x :: xs => f(x) >>= (a1 -> loop(xs, ks -> k(a1 :: ks)))
    };
    loop(l, upcast(Applicative.point))

~~~

`traverseM` would need to be an "interface" method in the Traversable class as it cannot be defined in terms of the existing members.

Fortunately the functions in `Foldable` that can benefit from short-circuiting are already monadic rather than applicative (`foldLeftM`, `foldRightM`, `foreachM`).

**Comment by stephentetley** [comment]
I will make a PR for this at some point. I've been busy recently updating libraries and code that I have for work so I haven't made the time to do it...

**Comment by magnus-madsen** [comment]
> I will make a PR for this at some point. I've been busy recently updating libraries and code that I have for work so I haven't made the time to do it...

No worries. We are thankful for all your help :)

**Comment by magnus-madsen** [comment]
@stephentetley Is this still relevant?

**Comment by stephentetley** [comment]
Yes, I think so.  We should have `sequenceM` and `forM` as well.
